Upgrade libmbedtls to version 2.27.0

xyzhao2018 opened the pull request:
Signed-off-by: xzhao3 <shirley.zhao@intel.com>

<!--
    If you are new to submitting pull requests to OP-TEE, then please have a
    look at the list below and tick them off before submitting the pull request.

    1. Read our contribution guidelines:
         documentation/github.md.

    2. Read the contribution section in Notice.md and pay extra attention to the
       "Developer Certificate of Origin" part:
         https://github.com/OP-TEE/optee_os/blob/master/Notice.md#contributions.

    3. You should run checkpatch preferably before submitting the pull request.

    4. When everything has been reviewed, you will need to squash, rebase and
       add tags like `Reviewed-by`, `Acked-by`, `Tested-by` etc.

    NOTE: This comment will not be shown in the pull request, so no harm keeping
    it, but feel free to remove it if you like.
-->

Commit:
lib: mbedtls: return TEE_ERROR_BAD_PARAMETERS on input data error

This change fixes Keymaster VTS if cryptolib uses libmedtls
EncryptionOperationsTest, RsaPkcs1Success and
EncryptionOperationsTest, RsaOaepSuccess probabilistic failure.
We should change error code from libmedtls to TEE_AsymmetricDecrypt.
In the same scenario, the tomcrypt return value is eventually
Converted to TEE_ERROR_BAD_PARAMETERS,and then pass the test.
But mbedtls converted to TEE_ERROR_BAD_STATE,
This causes TEE_AsymmetricDecrypt() to panic.

Signed-off-by: Liu Shiwei <liushiwei@eswin.com>
Tested-by: Liu Shiwei <liushiwei@eswin.com>
Acked-by: Etienne Carriere <etienne.carriere@linaro.org>
Acked-by: Jerome Forissier <jerome@forissier.org>
Acked-by: Victor Chong <victor.chong@linaro.org>
Acked-by: Jens Wiklander <jens.wiklander@linaro.org>

jforissier committed:
Squashed commit upgrading to mbedtls-2.27.0

Squash merging branch import/mbedtls-2.27.0

 335b95f50f90 ("core: libmbedtls: add ctr_drbg.c to core sources")
 9ad9df8b36e4 ("core: libtomcrypt: libmbedtls: mbedtls_mpi_montred() now returns void")
 ade0994c57b3 ("libmbedtls: add SM2 curve")
 3041cf9726e2 ("libmbedtls: mbedtls_mpi_exp_mod(): optimize mempool usage")
 a2e7a4cd262d ("libmbedtls: mbedtls_mpi_exp_mod(): reduce stack usage")
 87efbd27f8e0 ("libmbedtls: mbedtls_mpi_exp_mod() initialize W")
 e7c59b9b1d5f ("libmbedtls: fix no CRT issue")
 d76bd278d9e2 ("libmbedtls: add interfaces in mbedtls for context memory operation")
 e5b6c167f809 ("libmedtls: mpi_miller_rabin: increase count limit")
 b81d896a903d ("libmbedtls: add mbedtls_mpi_init_mempool()")
 3fbd8660c09d ("libmbedtls: make mbedtls_mpi_mont*() available")
 2cc759c67e37 ("mbedtls: configure mbedtls to reach for config")
 48bf81758c6e ("mbedtls: remove default include/mbedtls/config.h")
 3602df84d7b3 ("Import mbedtls-2.27.0")

Signed-off-by: Jerome Forissier <jerome@forissier.org>
Acked-by: Jens Wiklander <jens.wiklander@linaro.org>

GangSecurity commented:
you can port all the patches form 2.22.0 to 2.27.0, not just copy the code and make a patch.

jingdlu reviewed (commented):
How do you get all of these changes?

jingdlu reviewed (commented):
Two questions:
1, Is this original patch "Squashed commit upgrading to mbedtls-2.27.0" ?
2, I saw lots of changes regarding libtomcrypt. Do we still them?

jingdlu commented on core/lib/libtomcrypt/mpi_desc.c:
```diff
@@ -545,8 +545,7 @@ static int montgomery_reduce(void *a, void *b, void *c)
 
-	if (mbedtls_mpi_montred(&A, N, *mm, &T))
-		goto out;
+	mbedtls_mpi_montred(&A, N, *mm, &T);
```
This change should be in the commit "Squashed commit upgrading to mbedtls-2.27.0"? Right?

xyzhao2018 replied:
NO，Squashed commit upgrading to mbedtls-2.27.0 this patch is from CVE. The update above is not included in the patch.

jingdlu replied:
Please check this commit: https://github.com/OP-TEE/optee_os/commit/7901324d9530594155991c8b283023d567741cc7

jingdlu commented on lib/libmbedtls/mbedtls/library/ssl_tls.c:
```diff
@@ -187,6 +187,10 @@ int mbedtls_ssl_session_copy( mbedtls_ssl_session *dst,
     mbedtls_ssl_session_free( dst );
     memcpy( dst, src, sizeof( mbedtls_ssl_session ) );
 
+#if defined(MBEDTLS_SSL_SESSION_TICKETS) && defined(MBEDTLS_SSL_CLI_C)
```
all of this change come from CVE patch? If yes, please remove your signed-off-by.

xyzhao2018 replied:
No, this CVE patch can't be used directly, we patch it manually. So two names are remained.

jingdlu replied:
Then can we just change the code manually? I mean just append your signed-off-by and try to make the title shorter.

xyzhao2018 replied:
@GangSecurity , how to handle this case previously? Is there any precedents？

xyzhao2018 committed:
fix a session copy bug

fix a possible double reference on 'ticket'
when peer_cert/peer_cert_digest calloc failed.

Signed-off-by: 吴敬辉 <11137405@vivo.com>
Signed-off-by: xzhao3 <shirley.zhao@intel.com>
